Add support for pyexasol part 2/2

Nicoretti opened the pull request:
fixes #105

Nicoretti committed:
Add hooks for websocket based exasol dialect

Nicoretti committed:
Implement basic parsing of connection arguments

Nicoretti committed:
Fix parameterized queries

Nicoretti committed:
Add websocket dialect/connector to ci/cd matrix

Nicoretti committed:
Fix websocket dbapi signature and default implementation

Nicoretti committed:
Disable statement cache for websocket dialect

Nicoretti committed:
Fix CI/CD builds for websocket based dialect

Nicoretti committed:
Fix Big Integer conversion edge case for websocket dialect

Nicoretti committed:
Update dependencies

Nicoretti committed:
Restructure the dbapi2 module of the websocket driver

Nicoretti committed:
Move dbapi exceptions into seperate module

Nicoretti committed:
Move Connection and Cursor protocol definition different module

Nicoretti committed:
Move type definitions and type conversions to different module

Nicoretti committed:
Improve readability of section seperators

Nicoretti committed:
Move Connection and Curosr implementation in different modules

Nicoretti committed:
Move comment to correct location

kaklakariada commented on exasol/driver/websocket/_connection.py:
```diff
@@ -0,0 +1,144 @@
+        password: str = None,
+        schema: str = "",
+        autocommit: bool = True,
+        tls: bool = True,
```
Do we want to allow deactivating tls?
Idea for future improvement: also allow specifying the certificate fingerprint.

Nicoretti replied:
Not 100% sure yet, but wanted to stay compatible with pyexasol here for starters.
Having a "secure" default on top is fine imho.

kaklakariada commented on exasol/driver/websocket/_connection.py:
```diff
@@ -0,0 +1,144 @@
+            "http_proxy": None,
+            "client_name": None,
+            "client_version": None,
+            "client_os_username": None,
```
Add a note to allow specifying these values. Or will this be set by pyexasol?

Nicoretti replied:
this are all pyexasol options, available and it will fill most of them with it's default. I put it here for completeness
so if any parameters also needs to be exposed it is obvious and straight forward to do so.

kaklakariada commented on exasol/driver/websocket/_connection.py:
```diff
@@ -0,0 +1,144 @@
+        return DefaultCursor(self)
+
+    def __del__(self):
+        self.close()
```
Do we need error handling here?

Nicoretti replied:
No generally users should not depend on, or invoke  `__del__` directly. It generally is called by the interpreter and it's execution behavior may differ from that of other code ([see here](https://docs.python.org/3/reference/datamodel.html?highlight=__del__#object.__del__)).
This call rather was added as additional safety net, to prevent resource leakage.

kaklakariada commented on exasol/driver/websocket/_cursor.py:
```diff
@@ -0,0 +1,222 @@
+
+    @_is_not_closed
+    def callproc(self, procname, parameters=None):
+        """See also :py:meth: `Cursor.callproc`"""
+        raise NotSupportedError("Optional and therefore not supported")
```
Should we document this in the doc-string that this is not supported?

Nicoretti replied:
Lets talk about this one ;D

kaklakariada commented on exasol/driver/websocket/_cursor.py:
```diff
@@ -0,0 +1,222 @@
+            self.executemany(operation, [parameters])
+            return
+
+        self._cursor = connection.execute(operation)
```
```suggestion
        if parameters:
            self.executemany(operation, [parameters])
            return

        connection = self._connection.connection
        self._cursor = connection.execute(operation)
```

kaklakariada commented on exasol/driver/websocket/_cursor.py:
```diff
@@ -0,0 +1,222 @@
+            for params in seq_of_parameters
+        ]
+        connection = self._connection.connection
+        self._cursor = connection.cls_statement(connection, operation, prepare=True)
```
Why also pass `connection` as an argument?

tkilias replied:
This a bit mean cls_statement isn't a method or function, it is a class variable with type as value of the underlying pyexasol connection. Can you add a comment what is happening here.

kaklakariada commented on exasol/driver/websocket/_errors.py:
```diff
@@ -0,0 +1,76 @@
+"""
+
+
+class Warning(Exception):  # pylint: disable=W0622
```
Add a short justification why this warning is OK

kaklakariada commented on exasol/driver/websocket/_protocols.py:
```diff
@@ -0,0 +1,272 @@
+        Connect to the database.
+
+        Attention:
+            Addition not required by pep-249.
```
```suggestion
            Addition not required by PEP-249.
```

kaklakariada commented on exasol/driver/websocket/_protocols.py:
```diff
@@ -0,0 +1,272 @@
+    def cursor(self):
+        """
+        Return a new Cursor Object using the connection.
+
+        If the database does not provide a direct cursor concept, the module will have
+        to emulate cursors using other means to the extent needed
+        by this specification.
```
```suggestion
```
Does not apply

Nicoretti replied:
General "interface" description.

kaklakariada commented on exasol/driver/websocket/_protocols.py:
```diff
@@ -0,0 +1,272 @@
+        Commit any pending transaction to the database.
+
+        Note:
+            If the database supports an auto-commit feature, this must be initially off.
+            An interface method may be provided to turn it back on. Database modules
+            that do not support transactions should implement this method with
+            void functionality.
```
Exasol supports auto-commit -> adapt text

Nicoretti replied:
This is rather the dbapi2 spec/description, you have a point though and this should be addressed in the docs of the actual connection in `_connection.py`.

kaklakariada commented on exasol/driver/websocket/_protocols.py:
```diff
@@ -0,0 +1,272 @@
+
+    def rollback(self):
+        """
+        This method is optional since not all databases provide transaction support.
+
+        In case a database does provide transactions this method causes the database
+        to roll back to the start of any pending transaction. Closing a connection
+        without committing the changes first will cause an implicit rollback
+        to be performed.
```
```suggestion
        This method causes the database
        to roll back to the start of any pending transaction. Closing a connection
        without committing the changes first will cause an implicit rollback
        to be performed.
```

Nicoretti replied:
General "interface" description.

kaklakariada commented on exasol/driver/websocket/_protocols.py:
```diff
@@ -0,0 +1,272 @@
+        This read/write attribute specifies the number of rows to fetch
+        at a time with .fetchmany().
+
+        It defaults to 1 meaning to fetch a single row at a time.
```
```suggestion
        It defaults to 1, meaning to fetch a single row at a time.
```

kaklakariada commented on exasol/driver/websocket/_protocols.py:
```diff
@@ -0,0 +1,272 @@
+    def rowcount(self):
+        """
+        This read-only attribute specifies the number of rows that the last .execute*()
+        produced (for DQL statements like SELECT) or affected(for DML statements
```
```suggestion
        produced (for DQL statements like SELECT) or affected (for DML statements
```

kaklakariada commented on exasol/driver/websocket/_protocols.py:
```diff
@@ -0,0 +1,272 @@
+        like UPDATE or INSERT).
+
+        The attribute is -1 in case no .execute*() has been performed on the cursor or
+        the rowcount of the last operation is cannot be determined by the interface.
```
```suggestion
        the rowcount of the last operation cannot be determined by the interface.
```

kaklakariada commented on exasol/driver/websocket/_protocols.py:
General: add type hints to the protocols

kaklakariada commented on exasol/driver/websocket/_protocols.py:
```diff
@@ -0,0 +1,272 @@
+        (This method is optional since not all databases provide stored procedures)
+
+        The sequence of parameters must contain one entry for each argument that the
+        procedure expects. The result of the call is returned as modified copy of
```
```suggestion
        procedure expects. The result of the call is returned as a modified copy of
```

kaklakariada commented on exasol/driver/websocket/_protocols.py:
```diff
@@ -0,0 +1,272 @@
+        its behavior. This is most effective for algorithms where the same operation
+        is used, but different parameters are bound to it (many times).
+
+        For maximum efficiency when reusing an operation, it is best to usethe
```
```suggestion
        For maximum efficiency when reusing an operation, it is best to use the
```

kaklakariada commented on exasol/driver/websocket/_types.py:
```diff
@@ -0,0 +1,96 @@
+from time import localtime
+
+
+def Date(year: int, month: int, day: int) -> date:  # pylint: disable=C0103
```
Why use an upper-case name for a function?
And why do you need this function at all?

Nicoretti replied:
If you produce a type via a function (factory etc.) it sometimes provides a nicer and easier to grasp API usage. 

e.g.:

```python
date = Date(year=2022, month=12, day=1) # states more obviously that an object is being constructed
date = date(year=2022, month=12, day=1) # does the function has any side effects? e.g. setting the date?
``` 

That said, in this scenario the final answer to both of your questions is this:

> This is required by the `DBAPI2` spec

kaklakariada commented on exasol/driver/websocket/dbapi2.py:
```diff
@@ -0,0 +1,101 @@
+
+apilevel = "2.0"  # pylint: disable=C0103
+threadsafety = 1  # pylint: disable=C0103
+paramstyle = "qmark"  # pylint: disable=C0103
```
Justification for linter exclusion

Nicoretti replied:
The issue is the lower case names, they are required though by the DBAPI2 specification.
Thought that this should be "obvious" (for a python dev) and I didn't want to put more noise than necessary.
Your comment made me think though ;D.

kaklakariada commented on exasol/driver/websocket/dbapi2.py:
```diff
@@ -0,0 +1,101 @@
+    return connection.connect()
+
+
+__all__ = [
```
What does this do?

Nicoretti replied:
TL;DR if you do a `from <module> import *` those `symbols` get imported.
[This](https://docs.python.org/3/tutorial/modules.html?highlight=__all__#importing-from-a-package) explains it in terms of packages, the module behavior is pretty similar (except for the loading files part).

kaklakariada commented on setup.cfg:
```diff
@@ -1,6 +1,7 @@
 [db]
 exasol-pyodbc=exa+pyodbc://sys:exasol@localhost:8888/TEST?CONNECTIONCALL=en_US.UTF-8&DRIVER=EXAODBC&SSLCertificate=SSL_VERIFY_NONE
 exasol-turbodbc=exa+turbodbc://sys:exasol@localhost:8888/TEST?CONNECTIONCALL=en_US.UTF-8&DRIVER=EXAODBC&SSLCertificate=SSL_VERIFY_NONE
+exasol-websocket=exa+websocket://sys:exasol@localhost:8888/TEST?CONNECTIONCALL=en_US.UTF-8&DRIVER=EXAODBC&SSLCertificate=SSL_VERIFY_NONE
```
```suggestion
exasol-pyodbc=exa+pyodbc://sys:exasol@localhost:8563/TEST?CONNECTIONCALL=en_US.UTF-8&DRIVER=EXAODBC&SSLCertificate=SSL_VERIFY_NONE
exasol-turbodbc=exa+turbodbc://sys:exasol@localhost:8563/TEST?CONNECTIONCALL=en_US.UTF-8&DRIVER=EXAODBC&SSLCertificate=SSL_VERIFY_NONE
exasol-websocket=exa+websocket://sys:exasol@localhost:8563/TEST?CONNECTIONCALL=en_US.UTF-8&DRIVER=EXAODBC&SSLCertificate=SSL_VERIFY_NONE
```

Nicoretti replied:
The default port the exasol DB is setup with ITDE is 8888.

kaklakariada commented on sqlalchemy_exasol/websocket.py:
```diff
@@ -0,0 +1,68 @@
+
+class Integer(sqltypes.INTEGER):
+    def result_processor(self, dialect, coltype):
+        def to_integer(value):
+            if isinstance(value, str):
+                return int(value)
+            return value
+
+        return to_integer
```
Would be be shorter to return a `lambda`?

Nicoretti replied:
Shorter yes, more maintainable, I don't think so. Additionally  in the context of SQLA debugging can be
quite tricky, adding an lambda would not make it even harder.

```python
def result_processor(self, dialect, coltype)
    return lambda v: int(value) if isinstance(v, str) else v
```

kaklakariada commented on sqlalchemy_exasol/websocket.py:
```diff
@@ -0,0 +1,68 @@
+        def tls(value):
+            return True if not value == "SSL_VERIFY_NONE" else False
+
+        def autocommit(value):
+            value = value.lower()
+            mapping = defaultdict(
+                bool, {"y": True, "yes": "True", "n": False, "no": False}
+            )
+            return mapping[value]
```
Will this also work when `value` alread is a bool?

Nicoretti replied:
no the current implementation only works with arbitrary stings.

kaklakariada commented on sqlalchemy_exasol/websocket.py:
```diff
@@ -0,0 +1,68 @@
+        }
+        for name, value in known_options.items():
+            converter = converters[name]
+            kwargs[converter.name] = converter.map(value)
```
Where is `Converter.map()` defined?

Nicoretti replied:
At the top of the function, a named tuple (low cost type definition) is done.
In the lines 50-54, Two converter types are crated, e.g.:
```python
"SSLCertificate": Converter("tls", tls),
...
```

this also could have been written like this:
```python
"SSLCertificate": Converter(name="tls", map=tls),
...
````

Nicoretti committed:
Update exasol/driver/websocket/_protocols.py

Co-authored-by: Christoph Pirkl <christoph.pirkl@exasol.com>

Nicoretti committed:
Add todo note

Nicoretti committed:
Update exasol/driver/websocket/_protocols.py

Co-authored-by: Christoph Pirkl <christoph.pirkl@exasol.com>

Nicoretti committed:
Update exasol/driver/websocket/_protocols.py

Co-authored-by: Christoph Pirkl <christoph.pirkl@exasol.com>

Nicoretti committed:
Update exasol/driver/websocket/_protocols.py

Co-authored-by: Christoph Pirkl <christoph.pirkl@exasol.com>

Nicoretti committed:
Update exasol/driver/websocket/_protocols.py

Co-authored-by: Christoph Pirkl <christoph.pirkl@exasol.com>

Nicoretti committed:
Update exasol/driver/websocket/_protocols.py

Co-authored-by: Christoph Pirkl <christoph.pirkl@exasol.com>

Nicoretti committed:
Added justification for pylint suppressions

Nicoretti committed:
Add justification for linter suppression

Nicoretti committed:
Update exasol/driver/websocket/_cursor.py

Co-authored-by: Christoph Pirkl <christoph.pirkl@exasol.com>

Nicoretti committed:
Update dependencies

Nicoretti committed:
Fix sqlalchemy dialect conversions

Nicoretti committed:
Exasol Websocket DBAPI datatypes

Nicoretti committed:
Apply simplifications

Nicoretti committed:
Rework dbapi2 types

Nicoretti committed:
Add hooks for type conversions between pyexasol and dbapi2 cursor

Nicoretti committed:
Fix Date edge cases

Nicoretti committed:
Fix Datetime edge cases

redcatbear commented on doc/dataypes.rst:
```diff
@@ -0,0 +1,53 @@
+Questions:
+* Is the type Null a type? or a value or both?
+*
+
```
```suggestion

```

redcatbear commented on exasol/driver/websocket/_connection.py:
```diff
@@ -0,0 +1,146 @@
+
+class Connection:
+    """
+    Implementation of a websocket based connection.
```
```suggestion
    Implementation of a websocket-based connection.
```

redcatbear commented on exasol/driver/websocket/_connection.py:
```diff
@@ -0,0 +1,146 @@
+
+        Args:
+
+            dsn: Connection string, same format as standard JDBC/ODBC drivers.
```
```suggestion
            dsn: Connection string, same format as for standard JDBC / ODBC drivers.
```

redcatbear commented on exasol/driver/websocket/_connection.py:
```diff
@@ -0,0 +1,146 @@
+
+            dsn: Connection string, same format as standard JDBC/ODBC drivers.
+            username: which will be used for the authentication.
+            password: which will be used for the authentication
```
```suggestion
            password: which will be used for the authentication.
```

redcatbear commented on exasol/driver/websocket/_connection.py:
```diff
@@ -0,0 +1,146 @@
+
+    def close(self):
+        """See also :py:meth: `Connection.close`"""
+        if not self._connection:
```
why not invert the logic and remove the extra return?

redcatbear commented on exasol/driver/websocket/_connection.py:
```diff
@@ -0,0 +1,146 @@
+    def commit(self):
+        """See also :py:meth: `Connection.commit`"""
+        # TODO: document and make sure autocommit is turned of by default
+        #       see dbapi2 specification
```
```suggestion
        # TODO: document and make sure auto-commit is turned off by default
        #       see dbapi2 specification
```
Also, replace TODO with reference to ticket.

Nicoretti replied:
This is already documented in `_protocols.Connection`, i'll removed the TODO.

redcatbear commented on exasol/driver/websocket/_cursor.py:
```diff
@@ -0,0 +1,284 @@
+        "scale": "scale",
+        "display_size": "unknown",
+        "internal_size": "size",
+        "null_ok": "unknown",
```
Exasol supports the notion of nullables:
https://docs.exasol.com/db/latest/sql/desc[ribe].htm?Highlight=nullable#UsageNotes

Nicoretti replied:
Thanks for the link, as discussed we have a slightly different use case here.

redcatbear commented on exasol/driver/websocket/_cursor.py:
```diff
@@ -0,0 +1,284 @@
+
+def _is_not_closed(method):
+    """
+    Decorator requires the object to have a result.
```
The name of the decorator and the description seem only loosely connected.

tkilias replied:
yes, look like a copy paste error with belows decorator

Nicoretti replied:
I think it is a copy paste issue :grimacing:

Nicoretti replied:
Fixed it, thanks for pointing it out!

redcatbear commented on exasol/driver/websocket/_cursor.py:
```diff
@@ -0,0 +1,284 @@
+
+    def to_date(v):
+        if not isinstance(v, str):
+            return v
```
Shouldn't this check that the instance is a date before returning without change?

What if someone supplies a number.

tkilias replied:
then datetime.date.fromisoformat(v) would fail and this is probably a case where there is no other option then fail hard, because we got something from the DB which is garbage

Nicoretti replied:
As discussed this is a function scoped in a private function, my call sites are expected to be correct.

redcatbear commented on exasol/driver/websocket/_cursor.py:
```diff
@@ -0,0 +1,284 @@
+        """See also :py:meth: `Cursor.callproc`"""
+        raise NotSupportedError("Optional and therefore not supported")
+
+    @_is_not_closed
```
This hides double closing. Shouldn't there be a warning at least?

tkilias replied:
no, because it is a somewhat common thing, with try finally.

Nicoretti committed:
Add basic context manager functionality to default cursor

In order to be compatible with the exasol sqla base dialect the
a cursor needs to support the basic context manager functionality.
Therefore those have been added to the websocket based DBAPI2 cursor.

Nicoretti committed:
Added notes to context manager functions

Nicoretti committed:
Update dependencies

Nicoretti committed:
Add exception DBAPI 2 exception translation for executed statements

Nicoretti committed:
Skip test for duplicate names in result set for websocket based driver

Due to he fact that the current implementation of the websocket based
SQLA driver is based on PyExasol, duplicated names in the result set
aren't supported.

Nicoretti committed:
Fix DBAPI2 inconsistency for description attribute of the cursor

Nicoretti committed:
Remove datatypes section

Nicoretti committed:
Add notes regarding prepared statement issue

Nicoretti committed:
Fix unexpected auto commits (#336)

* Add regression test nox target
* Add regression test for invalid auto commit behavior (Issue-335)
* Add regression tests to integration test target
* Fix invalid auto commit behavior
* Update change-log
* Move odbc helpers to exasol driver(s) package

Co-authored-by: Sebastian Bär <redcatbear@ursus-minor.de>

Nicoretti committed:
Prepare release 4.4.0 (#337)


---------

Co-authored-by: Christoph Kuhnke <github@kuhnke.net>

Nicoretti committed:
Adjust date in change log (#338)

Nicoretti committed:
Update dependencies

Nicoretti committed:
Merge branch 'master' into features/#105-pyexasol-backend-part-2

Nicoretti committed:
Consolidate server version detection

Nicoretti committed:
Adjust test_update_executemany  test

Nicoretti committed:
Update dependencies

Nicoretti committed:
Delete implementation notes

Nicoretti committed:
Add support for enabeling and disabeling certificate validation

Nicoretti committed:
Fix tls settings in websocket driver

Nicoretti committed:
Make sure certificate tests works for all exasol dialects

Nicoretti committed:
Adjust DBAPI tests

Nicoretti committed:
Update README.md

* Add notes regarding wss beta status
* Add basic usage information regarding wss dialect

Nicoretti committed:
Update pylint rating

* Make sure lower bound is bumped too

Nicoretti committed:
Update dependencies

Nicoretti committed:
Fix rst formatting error

tkilias commented on README.rst:
```diff
@@ -139,3 +139,96 @@ Development & Testing
+
+.. attention::
+
+    The Websocket support currently is in Beta, therefore it should not be used in production.
```
```suggestion
    The Websocket support is currently in Beta, therefore it should not be used in production.
```

tkilias commented on README.rst:
```diff
@@ -139,3 +139,96 @@ Development & Testing
+.. attention::
+
+    The Websocket support currently is in Beta, therefore it should not be used in production.
+    We also recommend to have a look into the know issues, before you start using it.
```
```suggestion
    We also recommend to have a look into the known issues, before you start using it.
```

tkilias commented on README.rst:
```diff
@@ -139,3 +139,96 @@ Development & Testing
+    The Websocket support currently is in Beta, therefore it should not be used in production.
+    We also recommend to have a look into the know issues, before you start using it.
+
+    If you find encounter any issue, just `create an issue <https://github.com/exasol/sqlalchemy-exasol/issues/new?assignees=&labels=bug&projects=&template=bug.md&title=%F0%9F%90%9E+%3CInsert+Title%3E>`_.
```
```suggestion
    If you encounter any issue, please `create an issue <https://github.com/exasol/sqlalchemy-exasol/issues/new?assignees=&labels=bug&projects=&template=bug.md&title=%F0%9F%90%9E+%3CInsert+Title%3E>`_.
```

tkilias commented on README.rst:
```diff
@@ -139,3 +139,96 @@ Development & Testing
+    We also recommend to have a look into the know issues, before you start using it.
+
+    If you find encounter any issue, just `create an issue <https://github.com/exasol/sqlalchemy-exasol/issues/new?assignees=&labels=bug&projects=&template=bug.md&title=%F0%9F%90%9E+%3CInsert+Title%3E>`_.
+    With your feedback, we will be able stabilize this feature more quickly.
```
```suggestion
    With your feedback, we will be able to stabilize this feature more quickly.
```

tkilias commented on README.rst:
```````````````````````````diff
@@ -139,3 +139,96 @@ Development & Testing
+
+What is Websocket support?
+``````````````````````````
+In the context of SQLA and Exasol, websocket support means that an SQLA dialect
```````````````````````````
```suggestion
In the context of SQLA and Exasol, Websocket support means that an SQLA dialect
```

tkilias commented on README.rst:
```diff
@@ -139,3 +139,96 @@ Development & Testing
+
+* Less System Dependencies
+* Easier to use than ODBC based driver(s)
+* Lock free metadata calls etc.
```
This is also possible with pyodbc, I think

Nicoretti replied:
Yes, took it from the websocket protocol project. Not 100% sure about JDBC though.

tkilias commented on README.rst:
```diff
@@ -139,3 +139,96 @@ Development & Testing
+
+    # ATTENTION:
+    # In terms of security it is NEVER a good idea to turn of certificate validation!!
+    # In rare cases it may be handy for non security related reasons.
```
```suggestion
    # In rare cases it may be handy for non-security related reasons.
```

tkilias commented on README.rst:
```diff
@@ -139,3 +139,96 @@ Development & Testing
+    # In rare cases it may be handy for non security related reasons.
+    # That said, if you are not a 100% sure about your scenario, stick with the
+    # secure defaults.
+    # In most cases, having a valid certificate and/or configuring the truststore(s)
```
Should we add a link, how to todo this? Maybe also something for later.

Nicoretti replied:
We can, but this may something which already is or should be part of the TLS tutorial I think.

tkilias commented on README.rst:
```diff
@@ -139,3 +139,96 @@ Development & Testing
+* Various conversions regarding float, decimals
+    - Certain scenarios still yield a :code:`string` type instead :code:`float` or :code:`decimal` type.
+* Insert
+    - Insert Multiple does not work
```
is Insert Multiple a proper name?

Nicoretti replied:
I took this naming from the SQLA test suite, wasn't sure if it is SQLA terminology. Do you have a suggestion/alternative?

tkilias commented on README.rst:
```diff
@@ -139,3 +139,96 @@ Development & Testing
+    - Certain scenarios still yield a :code:`string` type instead :code:`float` or :code:`decimal` type.
+* Insert
+    - Insert Multiple does not work
+    - Insert from select does not work
```
```suggestion
    - Insert from SELECT does not work
```

tkilias commented on README.rst:
```diff
@@ -139,3 +139,96 @@ Development & Testing
+    - Insert from select does not work
+* For some prepared statements, the wss protocol type conversion does not work properly
+    - Error messages usually state some JSON type mismatch, e.g.: '... getString: JSON value is not a string ...'
+* Exists
```
You probabably mean the exists clause

Nicoretti committed:
Disable remaing failing SQLA compliance tests

See also:

* https://github.com/exasol/sqlalchemy-exasol/issues/341
* https://github.com/exasol/sqlalchemy-exasol/issues/342

tkilias commented on test/unit/exasol/test_websocket.py:
```diff
@@ -0,0 +1,95 @@
+                    "dsn": "localhost:8888",
+                    "password": "exasol",
+                    "username": "sys",
+                    "tls": True,
```
ok here is tls is ok, other connectors use for different versions of Exasol different encryption methods. https://docs.exasol.com/db/7.0/planning/data_security.htm?Highlight=encryption#GeneralConcepts

Nicoretti committed:
Update dependencies

tkilias commented on test/integration/exasol/test_update.py:
```diff
@@ -125,7 +125,9 @@ def test_update_executemany(self):
         result = testing.db.execute(stmt, values)
         expected = [(7, "jack2"), (8, "ed"), (9, "fred2"), (10, "chuck")]
-        assert result.rowcount == -1
+        # Depending on the dialect it either reports that the affected rows information
```
which one does what, maybe something which should be documented somewhere

Nicoretti committed:
Update exasol/driver/websocket/_connection.py

Co-authored-by: Sebastian Bär <redcatbear@ursus-minor.de>

Nicoretti committed:
Update exasol/driver/websocket/_connection.py

Co-authored-by: Sebastian Bär <redcatbear@ursus-minor.de>

Nicoretti committed:
Update exasol/driver/websocket/_connection.py

Co-authored-by: Sebastian Bär <redcatbear@ursus-minor.de>

Nicoretti committed:
Update README.rst

Co-authored-by: Torsten Kilias <tkilias@users.noreply.github.com>

Nicoretti committed:
Fix documentatio build

tkilias commented on test/integration/dbapi/conftest.py:
```diff
@@ -1,6 +1,6 @@
 import pytest
 
-from exasol.driver.websocket import connect
+from exasol.driver.websocket.dbapi2 import connect
```
doesn't the connetion fixture need a try finally to properly close the connection in case of errors

Nicoretti replied:
As discussed, we should be fine here.
* https://docs.pytest.org/en/7.1.x/how-to/fixtures.html#yield-fixtures-recommended
* https://docs.pytest.org/en/7.1.x/how-to/fixtures.html#handling-errors-for-yield-fixture

tkilias commented on sqlalchemy_exasol/websocket.py:
```diff
@@ -0,0 +1,152 @@
+        if not self.asdecimal:
+            return None
+
+        fstring = "%%.%df" % self._effective_decimal_return_scale
```
this might need an explanation, looks mean

Nicoretti replied:
Well, that's what I thought when I found and took it from the trubodbc dialect ;P

tkilias commented on exasol/driver/websocket/_connection.py:
```diff
@@ -0,0 +1,150 @@
+            "schema": schema,
+            "autocommit": autocommit,
+            "snapshot_transactions": None,
+            "connection_timeout": 10,
+            "socket_timeout": 30,
+            "query_timeout": 0,
+            "compression": False,
```
These parameter we might need to expose later, as well. They are sometimes necassary.             
            "connection_timeout": 10,
            "socket_timeout": 30,
            "query_timeout": 0,
            "compression": False,

tkilias commented on exasol/driver/websocket/_connection.py:
```diff
@@ -0,0 +1,150 @@
+            "quote_ident": False,
+            "json_lib": "json",
+            "verbose_error": True,
+            "debug": False,
```
some for the debug parameter

Nicoretti replied:
https://github.com/exasol/sqlalchemy-exasol/issues/341#issuecomment-1559383329

tkilias commented on exasol/driver/websocket/_connection.py:
```diff
@@ -0,0 +1,150 @@
+            "udf_output_connect_address": None,
+            "udf_output_dir": None,
+            "http_proxy": None,
+            "client_name": None,
```
We probably should set this, because this will be available in the DB logs later, and we probably need to distinct between pure pyexasol and sqlalchemy

Nicoretti replied:
good point

tkilias commented on exasol/driver/websocket/_connection.py:
```diff
@@ -0,0 +1,150 @@
+            "websocket_sslopt": {"cert_reqs": ssl.CERT_REQUIRED}
+            if certificate_validation
+            else None,
+            "access_token": None,
```
where these for SaaS?

Nicoretti replied:
AFIK yes (from a comment somewhere in pyexasol), besides that I have no idea  (yet) about that feature and it's usage.

tkilias commented on exasol/driver/websocket/_cursor.py:
```diff
@@ -0,0 +1,301 @@
+
+    name: str
+    type_code: TypeCode
+    display_size: Optional[int] = None
```
WTF, where does this come from

tkilias commented on exasol/driver/websocket/_cursor.py:
```diff
@@ -0,0 +1,301 @@
+    null_ok: Optional[bool] = None
+
+
+def _metadata(name, metadata) -> MetaData:
```
type hints

tkilias commented on exasol/driver/websocket/_cursor.py:
```diff
@@ -0,0 +1,301 @@
+    return value
+
+
+def _pyexasol2dbapi(value, metadata):
```
type hints

tkilias commented on exasol/driver/websocket/_cursor.py:
```diff
@@ -0,0 +1,301 @@
+
+               Difference:
+
+                    If the rowcount of the last operation cannot be determined it will
```
Rational? my guess Exasol behavior

Nicoretti replied:
Not sure if EXASOL imposes this behavior but, pyexasol does.

tkilias commented on exasol/driver/websocket/_cursor.py:
```diff
@@ -0,0 +1,301 @@
+        if rows is None:
+            return None
+
+        return tuple(self._convert_row(row) for row in rows)
```
is there a reason you generate a tuple instead of a list

tkilias commented on exasol/driver/websocket/_cursor.py:
```diff
@@ -0,0 +1,301 @@
+        except pyexasol.exceptions.ExaError as ex:
+            raise Error() from ex
+
+    def _convert(self, rows):
```
do we know the type of rows?

tkilias commented on exasol/driver/websocket/_cursor.py:
```diff
@@ -0,0 +1,301 @@
+
+        return tuple(self._convert_row(row) for row in rows)
+
+    def _convert_row(self, row):
```
do we know the type of row?

tkilias commented on exasol/driver/websocket/_cursor.py:
```diff
@@ -0,0 +1,301 @@
+
+    @_requires_result
+    @_is_not_closed
+    def fetchone(self):
```
type hints, or would this break with dbapi

Nicoretti replied:
I rather not, due to the fact that there 100% signature/typing in the pep.

tkilias commented on exasol/driver/websocket/_cursor.py:
```diff
@@ -0,0 +1,301 @@
+
+    @_requires_result
+    @_is_not_closed
+    def fetchmany(self, size=None):
```
type hints, or would this break with dbapi

Nicoretti replied:
I rather not, due to the fact that there 100% signature/typing in the pep.

tkilias commented on exasol/driver/websocket/_cursor.py:
```diff
@@ -0,0 +1,301 @@
+
+    @_requires_result
+    @_is_not_closed
+    def fetchall(self):
```
type hints, or would this break with dbapi

tkilias commented on exasol/driver/websocket/_cursor.py:
```diff
@@ -0,0 +1,301 @@
+        """
+
+    def __del__(self):
+        if self._is_closed:
```
set is_closed?

Nicoretti replied:
the `close` call will take care of that.

Nicoretti committed:
Restore old skipping behavior for pyodbc tests

Nicoretti committed:
Remove outdated TODO

Nicoretti committed:
Update README.rst

Co-authored-by: Torsten Kilias <tkilias@users.noreply.github.com>

Nicoretti committed:
Update README.rst

Co-authored-by: Torsten Kilias <tkilias@users.noreply.github.com>

Nicoretti committed:
Update README.rst

Co-authored-by: Torsten Kilias <tkilias@users.noreply.github.com>

Nicoretti committed:
Update README.rst

Co-authored-by: Torsten Kilias <tkilias@users.noreply.github.com>

Nicoretti committed:
Update changelog

Nicoretti committed:
Add rational for not following dbapi behavior in regard of the rowcount

Nicoretti committed:
Update README.rst

Co-authored-by: Torsten Kilias <tkilias@users.noreply.github.com>

Nicoretti committed:
Update README.rst

Co-authored-by: Torsten Kilias <tkilias@users.noreply.github.com>

Nicoretti committed:
Improve error message when accessing already closed cursor

Nicoretti committed:
Update README

Nicoretti committed:
Add client name for wss connections

Nicoretti committed:
Update README

Nicoretti committed:
Address code review feedback

Nicoretti committed:
Fix unit tests

Take new default parameter `client_name` into account.

tkilias commented on README.rst:
```diff
@@ -213,7 +213,7 @@ Known Issues
     - Certain scenarios still yield a :code:`string` type instead :code:`float` or :code:`decimal` type.
 * Insert
-    - Insert Multiple does not work
+    - Insert multiple entries via prepared statements does not work in all cases
```
```suggestion
    - Insert multiple rows via prepared statements does not work in all cases
```

tkilias commented on exasol/driver/websocket/_connection.py:
```diff
@@ -87,7 +90,7 @@ def __init__(
             "http_proxy": None,
-            "client_name": None,
+            "client_name": client_name,
             "client_version": None,
```
hm, maybe client version makes also sense and set the default to the version of this package

tkilias commented on exasol/driver/websocket/_connection.py:
```diff
@@ -87,7 +90,7 @@ def __init__(
+            "client_name": client_name,
             "client_version": None,
             "client_os_username": None,
             "protocol_version": 3,
```
Did you check if Exasol 7.1 also supports protocol version 3?

Nicoretti replied:
Implicitly, test in CI run against `7.1` and `7.0`

tkilias commented on CHANGELOG.rst:
```diff
@@ -3,6 +3,17 @@
+     **🚨 Attention:**
+
+        This feature is currently in Beta, therefore it should not be used in production.
+        We also recommend to have a look into the know issues, before you start using it.
```
```suggestion
        We also recommend to have a look into the known issues, before you start using it.
```

tkilias commented on CHANGELOG.rst:
```diff
@@ -3,6 +3,17 @@
+        This feature is currently in Beta, therefore it should not be used in production.
+        We also recommend to have a look into the know issues, before you start using it.
+
+        If you encounter any problem, just `create an issue <https://github.com/exasol/sqlalchemy-exasol/issues/new?assignees=&labels=bug&projects=&template=bug.md&title=%F0%9F%90%9E+%3CInsert+Title%3E>`_.
```
```suggestion
        If you encounter any problem, please `create an issue <https://github.com/exasol/sqlalchemy-exasol/issues/new?assignees=&labels=bug&projects=&template=bug.md&title=%F0%9F%90%9E+%3CInsert+Title%3E>`_.
```

Nicoretti committed:
Update CHANGELOG.rst

Co-authored-by: Torsten Kilias <tkilias@users.noreply.github.com>

Nicoretti committed:
Update CHANGELOG.rst

Co-authored-by: Torsten Kilias <tkilias@users.noreply.github.com>

Nicoretti committed:
Update README.rst

Co-authored-by: Torsten Kilias <tkilias@users.noreply.github.com>

Nicoretti committed:
Add client_version to connection parameters

exa-eswar commented on .github/workflows/ci-cd.yml:
```diff
@@ -64,6 +64,7 @@ jobs:
         connector:
           - pyodbc
           - turbodbc
+          - websocket
```
all ok
